### ISSUE TITLE
ENH: expose tool to translate mask from image space to raw data coordinates

### ISSUE THREAD

**Issue by silkenelson** [slac-lcls/smalldata_tools#34]
A user may have made a mask with an outside tool (pyfAI,....) which will be inn image space. psana works in raw coordinates. This translation lies inside of MakeMask in SmallDataAna_psana for e.g. the polygon or circle masks and could be exposed or added as an option (-i read mask-image or so)

**Comment by silkenelson** [comment]
There are two pieces of code in makeMask related to that:

From a circle:
                mask_router_nda = np.array( [(ix-cx)**2+(iy-cy)**2<ro**2 for ix, iy in zip(x,y)] )
                mask_rinner_nda = np.array( [(ix-cx)**2+(iy-cy)**2<ri**2 for ix, iy in zip(x,y)] )
                mask_r_nda = mask_router_nda&~mask_rinner_nda

From the selected points:
                p=np.array(ginput(nPoints))
                print(p)
                mpath=path.Path(p)
                all_p = np.array([ (ix,iy) for ix,iy in zip(x.flatten(),y.flatten()) ] )
                mask_r_nda = np.array([mpath.contains_points(all_p)]).reshape(x.shape)

**Comment by mcb64** [comment]
OK... let me see if I understand what is going on here.
In BaseSmallDataAna_psana.py, there is a function MakeMask.  This prompts for a shape: rectangle, circle, polygon, dark, noise, or edgepixels.  You want to also have an "image" option, which reads an image file, and translates it into a mask (an np.array of bool) and applies it to the detector in question?

**Comment by silkenelson** [comment]
Hm - first I was thinking an extra function, but I actually think that is not really clean. MakeMask currently prompts, so I think maybe add an optional parameter image=<nd_array> and if you pass that parameter, MakeMask only takes and does not prompt. It should still prompt for the saving part.

**Comment by mcb64** [comment]
So, currently creating a map is an iterative process.  You want to skip that, and set the mask to something based on the image parameter, if it is non-None.  Let it run through the "save to calibdir?" and "save to local?" bits at the end?

Or do you want it to continue to use the select loop, just use the image passed in as a first automatic selection, but giving the opportunity to select another option (if you select not done?)

**Comment by silkenelson** [comment]
I expect that users that create a mask' outside' use other tools to have to do what it needs to go, so that we can skip the iterative process. I'm not sure how to nearly ask for the array in the iterative process: a file-name & then we need to implement reading whatever (npz, npy, text, tiff, whatever pyFAI might write,...?) so I was thinking to skip that part and expect already an array. Then I'm not sure how to get that into the function in the loop, but I'm open to brilliant (or any) solutions rather than the cop-out I suggest.

**Comment by mcb64** [comment]
https://pyfai.readthedocs.io/en/0.21/man/pyFAI-drawmask.html seems to indicate that pyfai generates EDF files... so I think it is probably wise to just assume you are passed in a numpy array.  Then we just skip the loop entirely: if image is None: # existing loop code else: #anything we need to do with the array.  Then just do the last bits to save and possibly invert.

Do we assume we get a bool array of the proper shape, or do we want to reshape and compare to 0 ourselves?

**Comment by silkenelson** [comment]
So I expect we get an array of either 0/1 or bool (maybe we should display that). This array will e.g. be of 1672x1673 (or whatever image size the detector has, given the geometry-file/detector x/y) and we need to transform it into the raw data shape of tile*tile_x*tile_y and that array goes to the end where we display (?), maybe invert & save.
For completeness, we can also check the shape against the raw-data shape and simply save that in case someone does pass something like that in.

**Comment by mcb64** [comment]
The readability of this code is not enhanced by having tests like "dettype == 26" in it. :-)

I'm also a bit confused about mask_r_nda.  This seems to be the mask under development, which gets appended to the list of masks, mask, all of which are or-ed together to give totmask.  However, the size of mask_r_nda is a bit confusing:
    - For 'r' or 'R', this is created with code like:
           if needsGeo:
               mask_r_nda = np.array([CALCULATION for ix, iy in zip(ix, iy)])
           else:
               mask_r_nda = np.zeros_like(image)
    - For 'c', this is created like the 'needsGeo' calculation above.
    - For 'p', this is created with:
          mask_r_nda = np.array([calc(ix, iy) for ix, iy in zip(x.flatten(), y.flatten())])
          if self.__dict__[detname].det.dettype==30:
              mask_r_nda = np.array([mask_r_nda for itile in range(self.__dict__[detname].ped.shape[0])])
    - For 'd' or 'n', this is created with:
           mask_r_nda = np.zeros_like(pedResult)
       where pedResult is either det.pedestals(self.run) or det.rms(self.run).
    - For 'e' or 'cen', this is created with:
           mask_r_nda = np.zeros_like(det.coords_x(self.run))

These are, in the end, all supposed to be the same shape, right?  (And parenthetically, I'm not understanding the x vs ix distinction.)

**Comment by silkenelson** [comment]
Did I say the code was easy to read? Some consistency would help, this is rather historically developed under somewhat changing handling (or understanding?) or detector behavior and through the introduction of gain switching detectors where the pedestals stopped having the same shape as the raw data (and other new detectors before that also broke old assumptions (e.g. that a 2-d raw data shape would be no geometry was necessary - that is not so).
I'm a good 95% that this works for any detector one would reasonably create a mask for.

The upshot is if its too much work to read my code and/or just run that for some dataset e.g. in xpptut15, then we are likely better off leaving this untouched for now. And probably going through and understanding what is involved. is more effort than this feature is worth.
I personally am rather dismayed by the fact that I need to turn the image by 90 degree if I want the input values to make any sense, but failed to get this to work in a less hacky feeling way

**Comment by mcb64** [comment]
No, you didn't say it was easy.  And I don't think it's too much work, but if I understood it better, I could bring a little consistency to it perhaps.  (I'm totally used to this sort of "mudball" programming, when you have something that works, slap on a feature here, and another there... and before you know it, you have something large that *still* works, but is difficult to read and maintain.)

I've always found the easiest way to understand code is to ask someone who knows "What's going on here?"  If you don't have the time though, probably the thing to do is for me to take a look at the xpptut15 dataset and see what this does.

**Comment by silkenelson** [comment]
Let's talk next week how to best proceed: zoom with demo & discussion is likely best. I'm also not sure if it would not be better to move some of the code out into their own function: one of these uses is to go from the 'raw data space' to 'image space' (here, we have the psana methodand somewhat equivalent methods I have coded up) and _back_. I don't know how to do this in a generic way.

**Comment by mcb64** [comment]
So, I have this in two parts.  First, I have my existing pull request here, https://github.com/slac-lcls/smalldata_tools/pull/147, which changes all of the integer constants to symbolic names in the DetObjectClass and fixes a few other typos that I saw.

After this has been accepted, I have https://github.com/mcb64/smalldata_tools/tree/ENH/MakeExtMask, which is a further checkin which adds an extMask parameter to MakeMask, and uses this as the mask if it is the correct size and type.  Size and type checking is aided by the new rawShape parameter of the DetObject.  We test that this is set each time the data is read, setting it if needed.

Once the first pull request has been accepted, I will make another for this other branch.
